sql/src/main/java/io/crate/operation/projectors/DMLProjector.java

```diff
@@ -0,0 +1,144 @@
+import java.util.concurrent.atomic.AtomicReference;
+
+public abstract class DMLProjector<Request extends ShardRequest> extends AbstractProjector {
+
+    public static final int DEFAULT_BULK_SIZE = 1024;
+
+    private final AtomicReference<Throwable> upstreamFailure = new AtomicReference<>(null);
+
+    private final ShardId shardId;
+    private final CollectExpression<Row, ?> collectUidExpression;
+    private final Object lock = new Object();
+
+    protected final ClusterService clusterService;
+    protected final Settings settings;
+    protected final TransportActionProvider transportActionProvider;
+    protected final BulkRetryCoordinatorPool bulkRetryCoordinatorPool;
+    protected final UUID jobId;
+
+    protected BulkShardProcessor<Request> bulkShardProcessor;
+
+    public DMLProjector(ClusterService clusterService,
+                        Settings settings,
+                        ShardId shardId,
+                        TransportActionProvider transportActionProvider,
+                        BulkRetryCoordinatorPool bulkRetryCoordinatorPool,
+                        CollectExpression<Row, ?> collectUidExpression,
+                        UUID jobId) {
+        this.clusterService = clusterService;
+        this.settings = settings;
+        this.transportActionProvider = transportActionProvider;
+        this.bulkRetryCoordinatorPool = bulkRetryCoordinatorPool;
+        this.shardId = shardId;
+        this.collectUidExpression = collectUidExpression;
+        this.jobId = jobId;
+    }
+
+    protected abstract BulkShardProcessor<Request> createBulkShardProcessor(int bulkSize);
+
+    protected abstract ShardRequest.Item createItem(String id);
+
+    @Override
+    public void prepare(ExecutionState executionState) {
+        super.prepare(executionState);
+        bulkShardProcessor = createBulkShardProcessor(DEFAULT_BULK_SIZE);
+    }
+
+    @Override
+    public boolean setNextRow(Row row) {
+        final Uid uid;
+        synchronized (lock) {
+            // resolve the Uid
+            collectUidExpression.setNextRow(row);
+            uid = Uid.createUid(((BytesRef)collectUidExpression.value()).utf8ToString());
+        }
+        // routing is already resolved
+        bulkShardProcessor.addForExistingShard(shardId, createItem(uid.id()), null);
+        return true;
+    }
+
+    @Override
+    public void finish() {
+        bulkShardProcessor.close();
+        collectUpdateResultsAndPassOverRowCount();
+    }
+
+    @Override
+    public void fail(Throwable throwable) {
+        upstreamFailure.set(throwable);
+
+        if (throwable instanceof CancellationException) {
+            bulkShardProcessor.kill(throwable);
+        } else {
+            bulkShardProcessor.close();
+        }
+        collectUpdateResultsAndPassOverRowCount();
+    }
+
+    private void collectUpdateResultsAndPassOverRowCount() {
+        Futures.addCallback(bulkShardProcessor.result(), new FutureCallback<BitSet>() {
```

Comment:
would it be possible to re-use the BulkProcessorFutureCallback which is also used by the IndexWriterProjector?